azurelinuxagent/ga/collect_logs.py

```diff
@@ -0,0 +1,225 @@
+        return not self.should_run
+
+    def stop(self):
+        self.should_run = False
+        if self.is_alive():
+            self.join()
+
+    def init_protocols(self):
+        # The initialization of ProtocolUtil for the log collection thread should be done within the thread itself
+        # rather than initializing it in the ExtHandler thread. This is done to avoid any concurrency issues as each
+        # thread would now have its own ProtocolUtil object as per the SingletonPerThread model.
+        self.protocol_util = get_protocol_util()
+        self.protocol = self.protocol_util.get_protocol()
+
+    def daemon(self, init_data=False):
+        try:
+            if not self._log_collection_allowed():
+                return
+
+            if init_data:
+                self.init_protocols()
+
+            while not self.stopped():
+                try:
+                    for op in self._periodic_operations: # pylint: disable=C0103
+                        op.run()
+                except Exception as e: # pylint: disable=C0103
+                    logger.error("An error occurred in the log collection thread main loop; "
+                                 "will skip the current iteration.\n{0}", ustr(e))
+                finally:
+                    PeriodicOperation.sleep_until_next_operation(self._periodic_operations)
+        except Exception as e: # pylint: disable=C0103
+            logger.error("An error occurred in the log collection thread; will exit the thread.\n{0}", ustr(e))
+
+    def collect_and_send_logs(self):
+        if self._collect_logs():
+            self._send_logs()
+
+    @staticmethod
+    def _get_resource_limits():
+        # Define CPU limit (as percentage of CPU time) and memory limit (absolute value in megabytes).
+        cpu_limit = "5%"
+        memory_limit = "30M"  # K for kb, M for mb
+        return cpu_limit, memory_limit
+
+    @staticmethod
+    def _collect_logs():
+        logger.info("Starting log collection...")
+
+        # Invoke the command line tool in the agent to collect logs, with resource limits on CPU and memory (RAM).
+        scope_name = "collect-logs-{0}.scope".format(ustr(int(time.time() * 1000000)))
+        systemd_cmd = ["systemd-run", "--unit={0}".format(scope_name), "--scope"]
+
+        # More info on resource limits properties in systemd here:
+        # https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/resource_management_guide/sec-modifying_control_groups
+        cpu_limit, memory_limit = CollectLogsHandler._get_resource_limits()
+        resource_limits = ["--property=CPUAccounting=1", "--property=CPUQuota={0}".format(cpu_limit),
+                           "--property=MemoryAccounting=1", "--property=MemoryLimit={0}".format(memory_limit)]
+
+        # The log tool is invoked from the current agent's egg with the command line option
+        collect_logs_cmd = [get_python_cmd(), "-u", sys.argv[0], "-collect-logs"]
```

Comment:
Hmm, I used the same logic as the daemon when deciding to run the latest agent available, check out `update.py#run_latest`. Therefore `sys.executable` will always return the same interpreter as the `get_python_cmd` method since it also uses the platform one to start the extension handler.

Reply:
Update: I changed this to use `sys.executable` instead.